### PR TITLE
[PMG-3] Additional unittests

### PR DESCRIPTION
**Pull request by MGotkowski**
Tests added for code that was not tested.
pres_analyzer_coverage

### CODE REVIEW

File: src/presence_analyzer/tests.py [hunk 1]
```diff
@@ -53,6 +55,65 @@ def test_api_users(self):
...
+            ["Wed", 24465],
+            ["Thu", 23705],
+            ["Fri", 0],
+            ["Sat", 0],
+            ["Sun", 0]
+        ]
+        self.assertEqual(len(data), 7)
+        self.assertListEqual(data, result)
+
+    @patch("presence_analyzer.views.log")
```
Comment (tomaszszymanski129) [inline src/presence_analyzer/tests.py]: In STX, we're fans of single quotes. Replace all the doubles in this PR to single ones. About the mock itself - nicely done!

File: src/presence_analyzer/tests.py [hunk 2]
```diff
@@ -53,6 +55,65 @@ def test_api_users(self):
...
+
+    def test_api_presence_weekday(self):
+        """
+        Test presence time by weekday for one user
+        """
+        resp = self.client.get('/api/v1/presence_weekday/10')
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp.content_type, 'application/json')
+        data = json.loads(resp.data)
+        self.assertEqual(len(data), 8)
```
Comment (tomaszszymanski129) [inline src/presence_analyzer/tests.py]: later we compare data to a particular result to check if it equals, so we can safely remove this check.

File: src/presence_analyzer/tests.py [hunk 3]
```diff
@@ -85,6 +147,41 @@ def test_get_data(self):
             data[10][sample_date]['start'],
             datetime.time(9, 39, 5)
         )
+        self.assertTrue(mock_log.debug.called)
+
+    def test_mean(self):
+        """
+        Test mean() function.
+        """
+        mean = utils.mean
```
Comment (tomaszszymanski129) [inline src/presence_analyzer/tests.py]: no need for such reassignment, can be removed here and in subsequent tests as well.

File: src/presence_analyzer/tests.py [hunk 4]
```diff
@@ -85,6 +147,41 @@ def test_get_data(self):
...
+        start = datetime.time(11, 30, 0)
+        end = datetime.time(12, 30, 0)
+        self.assertEqual(3600, interval(start, end))
+
+    def test_seconds_since_midnight(self):
+        """
+        Test seconds_since_midnight() function.
+        """
+        calc_sec = utils.seconds_since_midnight
+        time = datetime.time(1, 1, 1)
```
Comment (tomaszszymanski129) [inline src/presence_analyzer/tests.py]: More edge cases please. At least two i can think of - (0, 0, 0) and (23, 59, 59). You don't need to add separate test for that, can be all in one.

File: src/presence_analyzer/tests.py [hunk 5]
```diff
@@ -85,6 +147,41 @@ def test_get_data(self):
...
+        self.assertEqual(3661, calc_sec(time))
+
+    def test_group_by_weekday(self):
+        """
+        Test assigning time to weekdays
+        """
+        group_by_weekday = utils.group_by_weekday
+        data = utils.get_data()
```
Comment (tomaszszymanski129) [inline src/presence_analyzer/tests.py]: Let's not make this dependent on a CSV of any sort.
Create an in-place test object and then pass it to grouping func.

File: runtime/data/test_data.csv [hunk 6]
```diff
@@ -6,4 +6,6 @@
 11,2013-09-10,09:19:50,13:55:54
 11,2013-09-11,09:13:26,16:15:27
 11,2013-09-12,10:18:36,16:41:25
-11,2013-09-13,13:16:56,15:04:02
\ No newline at end of file
+11,2013-09-13,13:16:56,15:04:02
+11,2013-09-13,13:16:56,150402
```
Comment (tomaszszymanski129) [inline runtime/data/test_data.csv]: On a second thought, let's keep this file consistent. Leave it untouched and add broken records on demand in a particular test.

File: src/presence_analyzer/tests.py [hunk 7]
```diff
@@ -85,6 +147,41 @@ def test_get_data(self):
...
+        self.assertEqual(0, mean([]))
+        self.assertEqual(3, mean([1, 2, 3, 4, 5]))
+
+    def test_interval(self):
+        """
+        Test interval() function.
+        """
+        interval = utils.interval
+        start = datetime.time(11, 30, 0)
+        end = datetime.time(12, 30, 0)
```
Comment (tomaszszymanski129) [inline src/presence_analyzer/tests.py]: More edge cases. Like start and end at the same time to see if that zeroes.

**Review by tomaszszymanski129 — CHANGES_REQUESTED** [review]
Not bad, however - coverage isn't only thing that matters, edge cases should be always included if you can think of any.

File: src/presence_analyzer/tests.py [hunk 8]
```diff
@@ -86,6 +149,78 @@ def test_get_data(self):
...
         )
 
+    @patch('presence_analyzer.utils.log')
+    def test_malformed_get_data(self, mock_log):
+        """
+        Test get_data() with malformed data
+        """
+        with patch.dict(main.app.config, {'DATA_CSV': MALF_DATA_CSV}):
+            utils.get_data()
+            self.assertTrue(mock_log.debug.called)
```
Comment (tomaszszymanski129) [inline src/presence_analyzer/tests.py]: assert itself can be moved out of with scope,
also, if we're sticking to decorators, we can move it like so:
image

**Review by tomaszszymanski129 — APPROVED** [review]
Nice, one minor suggestion.
